pdg/src/info.rs

```diff
@@ -67,18 +67,483 @@ fn set_flow_info(g: &mut Graph) {
+        let b1 = mk_copy(&mut g, b11);
+        let c11 = mk_field(&mut g, j, 0_u32);
+        let c1 = mk_copy(&mut g, c11);
+        let b2 = mk_store_addr(&mut g, b1);
+        let c2 = mk_store_addr(&mut g, c1);
+        let d1 = mk_field(&mut g, a, 1_u32);
+
+        let pdg = build_pdg(g);
+        assert!(info(&pdg, a).unique);
+        assert!(!info(&pdg, j).unique);
+        assert!(!info(&pdg, b1).unique);
+        assert!(!info(&pdg, b2).unique);
+        assert!(!info(&pdg, c1).unique);
+        assert!(!info(&pdg, c2).unique);
+        assert!(info(&pdg, d1).unique);
+    }
+
+    #[test]
+    fn field_vs_raw() {
+        let mut g = Graph::default();
+        //let mut a = Point {x: 0, y:0}; // A
+        //let b = &mut a;                // B1
+        //let c = &mut a.y;              // C1
+        //*b = 1;                        // B2
+        //*c = 2;                        // C2
+        //
+        //A
+        //|-------
+        //|      |
+        //B1     |y
+        //|      C1
+        //|      C2
+        //B2
+        let a = mk_addr_of_local(&mut g, 0_u32);
+        let b1 = mk_copy(&mut g, a);
+        let c11 = mk_field(&mut g, a, 1_u32);
+        let c1 = mk_copy(&mut g, c11);
+        let c2 = mk_store_addr(&mut g, c1);
+        let b2 = mk_store_addr(&mut g, b1);
+
+        let pdg = build_pdg(g);
+        assert!(!info(&pdg, a).unique);
+        assert!(!info(&pdg, b1).unique);
+        assert!(!info(&pdg, b2).unique);
+        assert!(!info(&pdg, c1).unique);
+        assert!(!info(&pdg, c2).unique);
+    }
+
+    #[test]
+    fn fields_different_levels() {
+        let mut g = Graph::default();
+        //let mut a = Point {x: 0, y:0}; // A
+        //let b = &mut a;                // B1
+        //let c = &mut b.y;              // C1
+        //let bb = &mut b.y;             // B2
+        //*b = 1;                        // B3
+        //*c = 2;                        // C
+        //
+        //A
+        //|-------
+        //|      |
+        //B1     |y
+        //|      C1
+        //|y
+        //B2
+        //
+        let a = mk_addr_of_local(&mut g, 0_u32);
+        let b1 = mk_copy(&mut g, a);
+        let c1 = mk_field(&mut g, a, 1_u32);
+        let b2 = mk_field(&mut g, b1, 1_u32);
+        let c2 = mk_store_addr(&mut g, c1);
+
+        let pdg = build_pdg(g);
+        assert!(!info(&pdg, a).unique);
+        assert!(!info(&pdg, b1).unique);
+        assert!(!info(&pdg, b2).unique);
+        assert!(!info(&pdg, c1).unique);
+        assert!(!info(&pdg, c2).unique);
+        assert!(!info(&pdg, b2).unique);
+    }
+
+    #[test]
+    fn lots_of_siblings() {
+        let mut g = Graph::default();
+        //let mut a = ColorPoint {x: 0, y: 0, z: Color{ r: 100, g: 100, b: 100}};   //A1
+        //let b = &mut a.x;                                                         //B1
+        //let c = &mut a.y;                                                         //C1
```

Comment:
thoughts on interleaving these comments above each respective line, rather than have them all at the top?

Reply:
I think it should be in doc comments for the function so that it's run and tested.  It's easy to make a small mistake as is.

Reply:
i don't see how these comments can be tested. considering that this function is the test itself, what do you envision?

Reply:
I mean we can test that they're well-formed Rust.

Reply:
I believe I have tested them all, but that would make it easier to be sure, and also if there's changes...

Reply:
my primary concern is attributing lines here to lines of test code below. i still would consider interleaving them for that purpose because i think having them spatially local helps with that. testing that it's well formed rust seems orthogonal to that